test/Chainweb/Test/Pact/PactInProcApi.hs

```diff
@@ -280,6 +281,156 @@ newBlockRewindValidate mpRefIO reqIO = testCase "newBlockRewindValidate" $ do
             $ mkExec' "(chain-data)"
       }
 
+pact420UpgradeTest :: TestBlockDb -> IO (IORef MemPoolAccess) -> WebPactExecutionService -> IO ()
+pact420UpgradeTest bdb mpRefIO pact = do
+
+  -- run past genesis, upgrades
+  forM_ [(1::Int)..3] $ \_i -> runCut'
+
+  -- run block 4
+  setMempool mpRefIO getBlock4
+  runCut'
+  pwo4 <- getPWO bdb cid
+
+  tx4_0 <- txResult 0 pwo4
+  assertEqual
+    "Should not resolve new pact natives"
+    (Just "Cannot resolve fold-db")
+    (tx4_0 ^? crResult . to _pactResult . _Left . to peDoc)
+
+  tx4_1 <- txResult 1 pwo4
+  assertEqual
+    "Should not resolve new pact natives"
+    (Just "Cannot resolve zip")
+    (tx4_1 ^? crResult . to _pactResult . _Left . to peDoc)
+
+  tx4_2 <- txResult 2 pwo4
+  assertEqual
+    "Load fdb module"
+    Nothing
+    (tx4_2 ^? crResult . to _pactResult . _Left . to peDoc)
+
+  cb4 <- cbResult pwo4
+  assertEqual "Coinbase events @ block 4" [] (_crEvents cb4)
+
+  -- run past v3 upgrade, pact 4 switch
+  setMempool mpRefIO mempty
+  cuts <- forM [(8::Int)..12] $ \_i -> do
+      runCut'
+      if _i == 10
+          then fmap Just (readMVar $ _bdbCut bdb)
+          else return Nothing
+
+  savedCut <- fromMaybeM (userError "A cut should exist here.") $ msum cuts
+
+  -- run block 22
+  setMempool mpRefIO $ getBlock13
+  runCut'
+  pwo13 <- getPWO bdb cid
+
+  cb13 <- cbResult pwo13
+  assertEqual "Coinbase events @ block 13" [] (_crEvents cb13)
+
+  tx13_0 <- txResult 0 pwo13
+  let m1 = PObject $ ObjectMap $ mempty
+        & M.insert (FieldKey "a") (PLiteral $ LInteger 1)
+        & M.insert (FieldKey "b") (PLiteral $ LInteger 1)
+      m2 = PObject $ ObjectMap $ mempty
+        & M.insert (FieldKey "a") (PLiteral $ LInteger 2)
+        & M.insert (FieldKey "b") (PLiteral $ LInteger 2)
+  assertEqual
+    "Should resolve fold-db pact native"
+    (Just $ PList $ V.fromList [m1,m2])
+    (tx13_0 ^? crResult . to _pactResult . _Right)
+
+  tx13_1 <- txResult 1 pwo13
+  assertEqual
+    "Should resolve zip pact native"
+    (Just $ PList $ V.fromList $ PLiteral . LInteger <$> [5,7,9])
+    (tx13_1 ^? crResult . to _pactResult . _Right)
+
+  -- rewind to savedCut (cut 18)
```

Comment:
we don't need this, so we also prob don't need to run so many blocks above, grab saved cut etc